include/velocypack/Iterator.h

```diff
@@ -49,123 +50,116 @@ class ArrayIterator {
 
-  ArrayIterator begin() const {
-    auto it = ArrayIterator(*this);
-    it._position = 0;
-    return it;
-  }
+  [[nodiscard]] ArrayIterator begin() const noexcept { return {*this}; }
 
-  ArrayIterator end() const {
-    auto it = ArrayIterator(*this);
+#if (defined(_MSC_VER) ||                                     \
+     (defined(__clang_version__) && __clang_major__ >= 12) || \
+     (!defined(__clang__) && defined(__GNUC__) && __GNUC__ >= 10))
+  [[nodiscard]] std::default_sentinel_t end() const noexcept { return {}; }
+  [[nodiscard]] bool operator==(std::default_sentinel_t) const noexcept {
+    return !valid();
+  }
+#else
+  [[nodiscard]] ArrayIterator end() const noexcept {
+    ArrayIterator it{*this};
     it._position = it._size;
     return it;
   }
+#endif
 
-  inline bool valid() const noexcept { return (_position < _size); }
+  [[nodiscard]] bool valid() const noexcept { return _position != _size; }
 
-  inline Slice value() const { return operator*(); }
+  [[nodiscard]] Slice value() const {
+    if (VELOCYPACK_UNLIKELY(!valid())) {
+      throw Exception{Exception::IndexOutOfBounds};
+    }
+    VELOCYPACK_ASSERT(_current != nullptr);
+    return Slice{_current};
+  }
 
-  inline void next() { operator++(); }
+  void next() noexcept {
+    if (VELOCYPACK_UNLIKELY(!valid())) {
+      return;
+    }
+    ++_position;
+    VELOCYPACK_ASSERT(_current != nullptr);
+    _current += Slice(_current).byteSize();
+  }
 
-  inline ValueLength index() const noexcept { return _position; }
+  [[nodiscard]] ValueLength index() const noexcept { return _position; }
 
-  inline ValueLength size() const noexcept { return _size; }
+  [[nodiscard]] ValueLength size() const noexcept { return _size; }
 
-  inline bool isFirst() const noexcept { return (_position == 0); }
+  [[deprecated(
+      "Dangerous API, you don't really know valid or not. "
+      "You can use index() and size().")]] bool
+  isFirst() const noexcept {
+    return _position == 0;
+  }
 
-  inline bool isLast() const noexcept { return (_position + 1 >= _size); }
+  [[deprecated(
+      "Dangerous API, you don't really know valid or not. "
+      "You can use index() and size().")]] bool
+  isLast() const noexcept {
+    return _position + 1 >= _size;
+  }
 
-  inline void forward(ValueLength count) {
-    if (_position + count >= _size) {
-      // beyond end of data
-      _current = nullptr;
+  void forward(ValueLength count) noexcept {
+    _position += count;
+    if (VELOCYPACK_UNLIKELY(_position >= _size)) {
       _position = _size;
-    } else {
-      auto h = _slice.head();
-      if (h == 0x13) {
-        while (count-- > 0) {
-          _current += Slice(_current).byteSize();
-          ++_position;
-        }
-      } else {
-        _position += count;
-        _current = _slice.at(_position).start();
+      return;
+    }
+    auto const head = _slice.head();
+    if (head == 0x13) {
+      while (count-- != 0) {
```

Comment:
This can still perform an out-of-bounds read (as before). I think we should limit the loop here so it doesn't read over the bounds.

Reply:
Why? We made check above

Reply:
```cpp
    _position += count;
    if (VELOCYPACK_UNLIKELY(_position >= _size)) {
      _position = _size;
      return;
    }
```

Reply:
Ah true, there was a `return` several lines below which I overlooked. All good then.